hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java

```diff
@@ -241,6 +243,22 @@ public SessionImpl(SessionFactoryImpl factory, SessionCreationOptions options) {
 			setHibernateFlushMode( initialMode );
 		}
 
+		if ( factory.getDefinedFilterNames().contains( TenantIdBinder.FILTER_NAME ) ) {
+			String tenantIdentifier = getTenantIdentifier();
+			if ( tenantIdentifier == null ) {
+				throw new HibernateException( "SessionFactory configured for multi-tenancy, but no tenant identifier specified" );
+			}
+			else {
+				CurrentTenantIdentifierResolver resolver = factory.getCurrentTenantIdentifierResolver();
+				if ( resolver==null || !resolver.isRoot(tenantIdentifier) ) {
+					// turn on the filter, unless this is the "root" tenant with access to all partitions
+					getLoadQueryInfluencers()
+							.enableFilter( TenantIdBinder.FILTER_NAME )
+							.setParameter( TenantIdBinder.PARAMETER_NAME, tenantIdentifier );
+				}
+			}
```

Comment:
What do you think about the following instead to support treating `null` as possible root?

```suggestion
			CurrentTenantIdentifierResolver resolver = factory.getCurrentTenantIdentifierResolver();
			String tenantIdentifier = getTenantIdentifier();
			if ( tenantIdentifier == null && ( resolver == null || !resolver.isRoot(tenantIdentifier) ) ) {
				throw new HibernateException( "SessionFactory configured for multi-tenancy, but no tenant identifier specified" );
			}
			else {
				if ( resolver==null || !resolver.isRoot(tenantIdentifier) ) {
					// turn on the filter, unless this is the "root" tenant with access to all partitions
					getLoadQueryInfluencers()
							.enableFilter( TenantIdBinder.FILTER_NAME )
							.setParameter( TenantIdBinder.PARAMETER_NAME, tenantIdentifier );
				}
			}
```